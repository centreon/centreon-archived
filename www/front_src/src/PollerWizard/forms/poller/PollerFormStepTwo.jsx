/* eslint-disable consistent-return */
/* eslint-disable array-callback-return */
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable react/prop-types */
/* eslint-disable react/prefer-stateless-function */
/* eslint-disable import/no-named-as-default */

import React, { Component } from 'react';

import { Field, reduxForm as connectForm } from 'redux-form';
import { withTranslation } from 'react-i18next';
import Select from 'react-select';

import { Typography, Button } from '@mui/material';

import styles from '../../../styles/partials/form/_form.scss';
import SelectField from '../../../components/form-fields/SelectField';
import CheckboxField from '../../../components/form-fields/CheckboxField';
import fieldHoc from '../../../components/form-fields/hoc';

class PollerFormStepTwo extends Component {
  state = {
    selectedAdditionals: [],
    selectedMaster: null,
  };

  /**
   * Get available additionals remote server
   * ==> all remote servers except selected master
   */
  getAvailableAdditionals = () => {
    const { pollers } = this.props;
    const { selectedMaster } = this.state;

    const availableAdditionals = pollers.filter((poller) => {
      if (poller.id !== selectedMaster) {
        return true;
      }
    });

    return availableAdditionals;
  };

  /**
   * Update selected master in state
   */
  handleChangeMaster = (event, value) => {
    const { change } = this.props;
    const { selectedAdditionals } = this.state;

    // remove selected additional if it's already the master
    const filteredAdditionals = value
      ? selectedAdditionals.filter((additional) => {
          if (additional.value !== value) {
            return true;
          }
        })
      : [];

    // update field value (mandatory cause it is connected to redux-form)
    change('linked_remote_slaves', filteredAdditionals);

    this.setState({
      selectedAdditionals: filteredAdditionals,
      selectedMaster: value,
    });
  };

  /**
   * Update selected additionals in state
   */
  handleChangeAdditionals = (event, values) => {
    this.setState({
      selectedAdditionals: values,
    });
  };

  render() {
    const { error, handleSubmit, onSubmit, pollers, t, goToPreviousStep } =
      this.props;
    const { selectedMaster } = this.state;

    const availableAdditionals = this.getAvailableAdditionals();

    return (
      <form autoComplete="off" onSubmit={handleSubmit(onSubmit)}>
        {pollers.length ? (
          <>
            <Typography variant="h6">
              {t('Attach poller to a master remote server')}
            </Typography>
            <Field
              component={SelectField}
              name="linked_remote_master"
              options={[
                {
                  text: '',
                  value: null,
                },
              ].concat(
                pollers.map((c) => ({
                  label: c.name,
                  text: c.name,
                  value: c.id,
                })),
              )}
              value={selectedMaster}
              onChange={this.handleChangeMaster}
            />
          </>
        ) : null}
        {selectedMaster && pollers.length >= 2 ? (
          <>
            <Typography variant="h6">
              {t('Attach poller to additional remote servers')}
            </Typography>
            <div className={styles['form-item']}>
              <Field
                isMulti
                component={fieldHoc(Select)}
                name="linked_remote_slaves"
                options={availableAdditionals.map((remote) => ({
                  label: remote.name,
                  value: remote.id,
                }))}
                onChange={this.handleChangeAdditionals}
              />
            </div>
          </>
        ) : null}
        <Field
          component={CheckboxField}
          defaultValue={false}
          label={t('Advanced: reverse Centreon Broker communication flow')}
          name="open_broker_flow"
        />
        <div className={styles['form-buttons']}>
          <Button size="small" onClick={goToPreviousStep}>
            {t('Previous')}
          </Button>
          <Button
            color="primary"
            size="small"
            type="submit"
            variant="contained"
          >
            {t('Apply')}
          </Button>
        </div>
        {error ? (
          <Typography color="error" variant="body2">
            {error.message}
          </Typography>
        ) : null}
      </form>
    );
  }
}

export default withTranslation()(
  connectForm({
    destroyOnUnmount: false,
    enableReinitialize: true,
    form: 'PollerFormStepTwo',
    keepDirtyOnReinitialize: true,
  })(PollerFormStepTwo),
);
