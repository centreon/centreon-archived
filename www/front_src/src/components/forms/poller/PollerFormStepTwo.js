import React, { Component } from "react";
import { Field, reduxForm as connectForm } from "redux-form";
import SelectField from "../../form-fields/SelectField";
import CheckboxField from "../../form-fields/CheckboxField";

class PollerFormStepTwo extends Component {
  render() {
    const { error, handleSubmit, onSubmit, pollers } = this.props;
    return (
      <div className="form-wrapper">
        <div className="form-inner">
          <div className="form-heading">
            <h2 className="form-title">Attach poller to a server</h2>
          </div>
          <form autocomplete="off" onSubmit={handleSubmit(onSubmit)}>
            {pollers ? (
              <Field
                name="linked_remote"
                component={SelectField}
                label="Select:"
                options={[
                  {
                    disabled: true,
                    selected: true,
                    text: "Select Remote Server",
                    value: ""
                  }
                ].concat(
                  pollers.map(c => ({ value: c.id, label: c.name, text:c.name }))
                )}
              />
            ) : null}
            <Field
              name="open_broker_flow"
              component={CheckboxField}
              label="Advanced: reverse Centreon Broker communication flow"
            />
            <div class="form-buttons">
              <button className="button" type="submit">
                Apply
              </button>
            </div>
            {error ? <div class="error-block">{error.message}</div> : null}
          </form>
        </div>
      </div>
    );
  }
}

const validate = () => ({});

export default connectForm({
  form: "PollerFormStepTwo",
  validate,
  warn: () => {},
  enableReinitialize: true
})(PollerFormStepTwo);
