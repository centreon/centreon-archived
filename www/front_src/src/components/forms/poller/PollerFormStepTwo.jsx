/* eslint-disable consistent-return */
/* eslint-disable array-callback-return */
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable react/prop-types */
/* eslint-disable react/prefer-stateless-function */
/* eslint-disable import/no-named-as-default */

import React, { Component } from 'react';

import { Field, reduxForm as connectForm } from 'redux-form';
import { withTranslation } from 'react-i18next';
import Select from 'react-select';

import styles from '../../../styles/partials/form/_form.scss';
import SelectField from '../../form-fields/SelectField';
import CheckboxField from '../../form-fields/CheckboxField';
import fieldHoc from '../../form-fields/hoc';

class PollerFormStepTwo extends Component {
  state = {
    selectedMaster: null,
    selectedAdditionals: [],
  };

  /**
   * Get available additionals remote server
   * ==> all remote servers except selected master
   */
  getAvailableAdditionals = () => {
    const { pollers } = this.props;
    const { selectedMaster } = this.state;

    const availableAdditionals = pollers.filter((poller) => {
      if (poller.id !== selectedMaster) {
        return true;
      }
    });

    return availableAdditionals;
  };

  /**
   * Update selected master in state
   */
  handleChangeMaster = (event, value) => {
    const { change } = this.props;
    const { selectedAdditionals } = this.state;

    // remove selected additional if it's already the master
    const filteredAdditionals = value
      ? selectedAdditionals.filter((additional) => {
          if (additional.value !== value) {
            return true;
          }
        })
      : [];

    // update field value (mandatory cause it is connected to redux-form)
    change('linked_remote_slaves', filteredAdditionals);

    this.setState({
      selectedMaster: value,
      selectedAdditionals: filteredAdditionals,
    });
  };

  /**
   * Update selected additionals in state
   */
  handleChangeAdditionals = (event, values) => {
    this.setState({
      selectedAdditionals: values,
    });
  };

  render() {
    const { error, handleSubmit, onSubmit, pollers, t } = this.props;
    const { selectedMaster } = this.state;

    const availableAdditionals = this.getAvailableAdditionals();

    return (
      <div className={styles['form-wrapper']}>
        <div className={styles['form-inner']}>
          <form autoComplete="off" onSubmit={handleSubmit(onSubmit)}>
            {pollers.length ? (
              <>
                <h2 className={styles['form-title']}>
                  {t('Attach poller to a master remote server')}
                </h2>
                <Field
                  name="linked_remote_master"
                  component={SelectField}
                  options={[
                    {
                      text: '',
                      value: null,
                    },
                  ].concat(
                    pollers.map((c) => ({
                      value: c.id,
                      label: c.name,
                      text: c.name,
                    })),
                  )}
                  value={selectedMaster}
                  onChange={this.handleChangeMaster}
                />
              </>
            ) : null}
            {selectedMaster && pollers.length >= 2 ? (
              <>
                <h2 className={styles['form-title']}>
                  {t('Attach poller to additional remote servers')}
                </h2>
                <div className={styles['form-item']}>
                  <Field
                    name="linked_remote_slaves"
                    component={fieldHoc(Select)}
                    options={availableAdditionals.map((remote) => ({
                      value: remote.id,
                      label: remote.name,
                    }))}
                    isMulti
                    onChange={this.handleChangeAdditionals}
                  />
                </div>
              </>
            ) : null}
            <Field
              name="open_broker_flow"
              component={CheckboxField}
              label={t('Advanced: reverse Centreon Broker communication flow')}
              defaultValue={false}
            />
            <div className={styles['form-buttons']}>
              <button className={styles.button} type="submit">
                {t('Apply')}
              </button>
            </div>
            {error ? (
              <div className={styles['error-block']}>{error.message}</div>
            ) : null}
          </form>
        </div>
      </div>
    );
  }
}

export default withTranslation()(
  connectForm({
    form: 'PollerFormStepTwo',
    enableReinitialize: true,
    destroyOnUnmount: false,
    keepDirtyOnReinitialize: true,
  })(PollerFormStepTwo),
);
