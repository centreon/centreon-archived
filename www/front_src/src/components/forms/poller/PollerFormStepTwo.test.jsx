import '@babel/polyfill';
import React from 'react';

import configureStore from 'redux-mock-store';
import { render, fireEvent, queryByAttribute } from '@testing-library/react';
import { Provider } from 'react-redux';

import PollerFormStepTwo from './PollerFormStepTwo';

const queryByName = queryByAttribute.bind(null, 'name');

const mockStore = configureStore([]);

const NoPoller = [];
const OnePoller = [
  {
    id: 1,
    name: 'poller1',
  },
];
const TwoPollers = [
  {
    id: 1,
    name: 'poller1',
  },
  {
    id: 2,
    name: 'poller2',
  },
];

// Initialize mockstore with empty state
const store = mockStore({});

describe('PollerFormStepTwo', () => {
  it('does not display selects if no one remote server exists', () => {
    const { container } = render(
      <Provider store={store}>
        <PollerFormStepTwo pollers={NoPoller} onSubmit={jest.fn()} />
      </Provider>,
    );

    expect(queryByName(container, 'linked_remote_master')).toBeNull();
    expect(queryByName(container, 'linked_remote_slaves')).toBeNull();
  });

  it('displays only master select if one remote server exists', () => {
    const { container } = render(
      <Provider store={store}>
        <PollerFormStepTwo pollers={OnePoller} onSubmit={jest.fn()} />
      </Provider>,
    );

    expect(queryByName(container, 'linked_remote_master')).toBeInTheDocument();
    expect(queryByName(container, 'linked_remote_slaves')).toBeNull();
  });

  it('displays slaves select if master is selected and several remotes exist', async () => {
    const { container } = render(
      <Provider store={store}>
        <PollerFormStepTwo pollers={TwoPollers} onSubmit={jest.fn()} />
      </Provider>,
    );

    const masterSelect = queryByName(container, 'linked_remote_master');
    expect(masterSelect).toBeInTheDocument();

    fireEvent.change(masterSelect, { target: { value: '1' } });

    expect(queryByName(container, 'linked_remote_slaves')).toBeInTheDocument();
  });
});
