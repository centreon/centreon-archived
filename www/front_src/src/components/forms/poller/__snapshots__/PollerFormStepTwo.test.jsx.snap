// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollerFormStepTwo renders 1`] = `
<div
  class="form-wrapper"
>
  <div
    class="form-inner"
  >
    <form
      autocomplete="off"
    >
      <h2
        class="form-title"
      >
        <span>
          Attach poller to a master remote server
        </span>
      </h2>
      <div
        class="form-group select"
      >
        <select
          class="form-control custom-select"
          id="field-linked_remote_master-1"
          name="linked_remote_master"
        >
          <option />
          <option
            label="poller1"
            value="1"
          >
            poller1
          </option>
          <option
            label="poller2"
            value="2"
          >
            poller2
          </option>
        </select>
      </div>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox orange"
        >
          <input
            class="custom-control-input"
            id="field-open_broker_flow-2"
            name="open_broker_flow"
            type="checkbox"
            value="false"
          />
          <label
            class="custom-control-label"
            for="field-open_broker_flow-2"
          >
            Advanced: reverse Centreon Broker communication flow
          </label>
        </div>
      </div>
      <div
        class="form-buttons"
      >
        <button
          class="button"
          type="submit"
        >
          <span>
            Apply
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
